src/optimizer/expr_simplification.ts

```diff
@@ -0,0 +1,713 @@
+        // building the AST later.
+        // Since receivers do not have names, I will use their id in their original ast
+        // as key.
+        const newReceiversMap: Map<number, ReceiverDescription> = new Map();
+
+        for (const r of t.receivers) {
+            const statementsResult = process_statements(
+                r.ast.statements,
+                ctx,
+                interpreter,
+            );
+            const newStatements = statementsResult.stmts;
+            ctx = statementsResult.ctx;
+            const newReceiverCode = cloneAstNode({
+                ...r.ast,
+                statements: newStatements,
+            });
+            const newReceiverDescription = { ...r, ast: newReceiverCode };
+            newReceivers.push(newReceiverDescription);
+            newReceiversMap.set(r.ast.id, newReceiverDescription);
+        }
+
+        // Process methods
+
+        // This is already a map in TypeDescription. This is the reason
+        // I did not need a separate map, like in the previous cases.
+        const newMethods: Map<string, FunctionDescription> = new Map();
+
+        for (const [name, m] of t.functions) {
+            if (m.ast.kind === "function_def") {
+                const statementsResult = process_statements(
+                    m.ast.statements,
+                    ctx,
+                    interpreter,
+                );
+                const newStatements = statementsResult.stmts;
+                ctx = statementsResult.ctx;
+                const newMethodCode = cloneAstNode({
+                    ...m.ast,
+                    statements: newStatements,
+                });
+                newMethods.set(name, { ...m, ast: newMethodCode });
+            } else {
+                // The rest of kinds do not have explicit Tact expressions, so just copy the current function description
+                newMethods.set(name, m);
+            }
+        }
+
+        // Now, we need to create the new AST, depending on its kind.
+        let newAst: AstTypeDecl;
+
+        switch (t.ast.kind) {
+            case "primitive_type_decl": {
+                newAst = t.ast;
+                break;
+            }
+            case "struct_decl":
+            case "message_decl": {
+                newAst = cloneAstNode({
+                    ...t.ast,
+                    fields: newFields.map((f) => f.ast),
+                });
+                break;
+            }
+            case "trait": {
+                const newDeclarations: AstTraitDeclaration[] = [];
+
+                for (const decl of t.ast.declarations) {
+                    switch (decl.kind) {
+                        case "asm_function_def":
+                        case "function_decl":
+                        case "function_def": {
+                            const newCode = newMethods.get(idText(decl.name))!
+                                .ast as AstTraitDeclaration;
```

Comment:
Both `!` and `as` are unsafe. Either the type should be fixed, or there should be a condition with `throwInternalCompilerError` here.